src/node.rs

```diff
@@ -2284,13 +2324,93 @@ impl<S: Send + Sync + 'static + ForkSource + std::fmt::Debug> EthNamespaceT for
         })
     }
 
+    /// Returns the value from a storage position at a given address.
+    ///
+    /// # Arguments
+    ///
+    /// * `address`: Address of the storage
+    /// * `idx`: Integer of the position in the storage
+    /// * `block`: The block storage to target
+    ///
+    /// # Returns
+    ///
+    /// A `BoxFuture` containing a `jsonrpc_core::Result` that resolves to a [H256] value in the storage.
     fn get_storage(
         &self,
-        _address: zksync_basic_types::Address,
-        _idx: U256,
-        _block: Option<zksync_types::api::BlockIdVariant>,
+        address: zksync_basic_types::Address,
+        idx: U256,
+        block: Option<zksync_types::api::BlockIdVariant>,
     ) -> jsonrpc_core::BoxFuture<jsonrpc_core::Result<zksync_basic_types::H256>> {
-        not_implemented("get_storage")
+        let inner = Arc::clone(&self.inner);
+
+        Box::pin(async move {
+            let mut reader = match inner.write() {
+                Ok(r) => r,
+                Err(_) => {
+                    return Err(into_jsrpc_error(Web3Error::InternalError));
+                }
+            };
+
+            let storage_key = StorageKey::new(AccountTreeId::new(address), u256_to_h256(idx));
+
+            let block_number = block
+                .map(|block| match block {
+                    zksync_types::api::BlockIdVariant::BlockNumber(block_number) => {
+                        Ok(utils::to_real_block_number(
+                            block_number,
+                            U64::from(reader.current_miniblock),
+                        ))
+                    }
+                    zksync_types::api::BlockIdVariant::BlockNumberObject(o) => {
+                        Ok(utils::to_real_block_number(
+                            o.block_number,
+                            U64::from(reader.current_miniblock),
+                        ))
+                    }
+                    zksync_types::api::BlockIdVariant::BlockHashObject(o) => reader
+                        .blocks
+                        .get(&o.block_hash)
+                        .map(|block| block.number)
+                        .ok_or_else(|| {
+                            log::error!("unable to map block number to hash #{:#x}", o.block_hash);
+                            into_jsrpc_error(Web3Error::InternalError)
+                        }),
+                })
+                .unwrap_or_else(|| Ok(U64::from(reader.current_miniblock)))?;
+
+            if block_number.as_u64() == reader.current_miniblock {
+                Ok(H256(reader.fork_storage.read_value(&storage_key).0))
+            } else if reader.block_hashes.contains_key(&block_number.as_u64()) {
+                reader
+                    .block_hashes
+                    .get(&block_number.as_u64())
+                    .and_then(|block_hash| reader.previous_states.get(block_hash))
+                    .and_then(|state| state.get(&storage_key))
```

Comment:
I see two potential bugs here:

Consider this scenario: we forked testnet at block 1000, and did a transaction in block 1001 that did:
* insert(0x1, 0x1)
* delete(0x2)


Now the bugs:
* if we query for key (0x3) at block 1001 - we'll get empty value (as this key is not present in state - as we didn't modify it) - but in practice, we should query the testnet at block 1000 (! - important - at fork level) to get the current value of the key there

* the other bug is about deletions - and it might exist in the current system too (but we'll have to check). When we query at block 1001 for key 0x2 (that we just deleted) - we should return 0x0 then (as the value was deleted) - but from what I see, we'll query the testnet and might return whatever value was set there before..

let's make sure that we have test coverage for these corner cases.

Reply:
I was under the assumption that forking at block 1000 would copy all the state of the block `1000`. So when the state would be archived/cloned it would still persist the keys that weren't modified. If that's not the case then we could effectively query the fork if the state entry isn't found. Any blocks created later would always have the new state additions in the cloned states.

As for the deletion, I currently see the `set_value` function in `StorageView` which leads me to believe (correct me if I'm wrong) that a deletion is achieved via setting the storage to `H256::zero()` so if a state entry `0x2` is removed via setting it to `H256::zero()` that would still be tracked, isn't it?

Reply:
I don't believe we're copying the full state of the block because in `from_url_and_miniblock_and_chain`, we're only fetching `get_block_details` and `get_block_by_hash`. The only information we're saving from the fork on start-up is:
```rust
        ForkDetails {
            fork_source: HttpForkSource::new(url.to_owned(), cache_config),
            l1_block: l1_batch_number,
            l2_block: block,
            block_timestamp: block_details.base.timestamp,
            l2_miniblock: miniblock,
            l2_miniblock_hash: root_hash,
            overwrite_chain_id: chain_id,
            l1_gas_price: block_details.base.l1_gas_price,
        }
```

I can dig into the deletion logic

Reply:
Yeah you're right on the copying block state. Upon further discovery I realized the mistake of my presumption and added the fix. Still looking into deletion.

Reply:
@mm-zk do you have some additional context on the delete? Looking into `zksync-era` repo I was unable to find any references where storage is deleted. As far as solidity smart contracts are concerned, `delete` merely [assigns](https://docs.soliditylang.org/en/develop/types.html#delete) the zero-value of the type to a key. Thus the deleted key must be present in `modified_keys` and subsequently set to 0.